test/Serialization/clang-function-types.swift

```diff
@@ -0,0 +1,31 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -emit-module -o %t %S/Inputs/def_clang_function_types.swift -use-clang-function-types
+// RUN: llvm-bcanalyzer %t/def_clang_function_types.swiftmodule | %FileCheck -check-prefix=CHECK-BCANALYZER %s
+// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -use-clang-function-types -experimental-print-full-convention -emit-sil -sil-debug-serialization -I %t %s -O | %FileCheck %s
+
+import def_clang_function_types
+
+// CHECK-BCANALYZER-LABEL: (INDEX_BLOCK):
+// CHECK-BCANALYZER: CLANG_TYPE_OFFSETS
+
+// CHECK-LABEL: sil hidden @$s4main5test1yyF
+func test1() {
+  // FIXME: this mangling will have to change
+  // CHECK: global_addr @$s24def_clang_function_types11has_fp_types13OpaquePointerVSgyXCSgvp : $*Optional<@convention(c) () -> Optional<OpaquePointer>>
+  let fp = has_fp_type
+  _ = fp?()
+}
+// CHECK-LABEL: } // end sil function '$s4main5test1yyF'
+
+// CHECK-LABEL: sil hidden @$s4main5test2yyF
+func test2() {
+  use_fp_internally()
+}
+// CHECK-LABEL: } // end sil function '$s4main5test2yyF'
+
+// CHECK-LABEL: sil public_external [canonical] @$s24def_clang_function_types17use_fp_internallyyyF
+// CHECK:         enum $Optional<@convention(c) () -> Optional<OpaquePointer>>, #Optional.none!enumelt
+// CHECK:         [[FN:%.*]] = function_ref @$s24def_clang_function_types9use_inout3argyxz_tlF : $@convention(thin) <τ_0_0> (@inout τ_0_0) -> ()
+// FIXME: something is very wrong in the printer here
```

Comment:
Is this something that is temporarily present for while you're working on the tests/making CI pass? What is going wrong here?

Reply:
I didn't change anything about the printer, so I don't know what's going on.

Reply:
I guess my question is: what is the symptom you are seeing? I couldn't understand that from the test case text. I understand you don't have a diagnosis/solution because you didn't fiddle with the printer.

Reply:
Long-standing bug, apparently.

Reply:
The bug was that it was printing `Optional<@convention(c) () -> Optional<SomePointer>>` as `@convention(c) () -> SomePointer??`.